### ISSUE TITLE
cleaning registry crates cache

### ISSUE THREAD

**Issue by y0y0z** [matthiaskrgr/cargo-cache#72]
i am just leaning rust and was wondering what are the tools for dependency management. i came across `cargo-cache`.

i am using cargo-cache version 0.3.2
```bash
$ cargo install --list | grep cache | head -n1
cargo-cache v0.3.2:
```

i have tried any combination of `cargo cache`  to clean cached crates. but when i look at `~/.cargo/registry/cache/*/` i still see the crates there.

1. if you are aware of it, would you please explain why it is not being cleaned?
2. making sure i didn't miss anything, is there any `cache` command that will achieve cleaning the `~/.cargo/registry/cache/*/` path? (if there is no such one, i will try to contribute; but as i am a new rustacean so it might take me time)

**Comment by matthiaskrgr** [comment]
What command did you run to clean the cache?

**Comment by y0y0z** [comment]
@matthiaskrgr:
```bash
$ ls -x1 ~/.cargo/registry/cache/*/ | wc -l
      11
$ cargo cache -a
$ cargo cache -e
$ cargo cache -g
ls -x1 ~/.cargo/registry/cache/*/ | wc -l
      11
```

as you can see, the amount of files remained after running `crago cache`

**Comment by matthiaskrgr** [comment]
Hmm ok, that should work I think.
What kind of files are left over by cargo-cache?

**Comment by y0y0z** [comment]
@matthiaskrgr: all the cached crates :)

**Comment by matthiaskrgr** [comment]
Hm so it's just regular .crates?
That sounds strange...
Can you check if your user is perhaps lacking permission to edit/remove these files?

**Comment by matthiaskrgr** [comment]
Looking at the code, there should actually be warnings if cargo-cache fails to remove a file so that might not be it.

**Comment by y0y0z** [comment]
@matthiaskrgr: my user does have permission on these files. i am able to remove it manually via
```bash
$ rm -rf ~/.cargo/registry/cache/*/
```

**Comment by matthiaskrgr** [comment]
Ok, turns out that `-a`, `-e` or `-g` (note that `-e` is essentially `-a` + `-g`) does not clear the `.crate` sources by design. :) 
I think you want something like `--remove-dir all` ?

note: I just found a related crash, looking into it...

**Comment by y0y0z** [comment]
@matthiaskrgr: i uninstalled cargo-cache and installed it again to get your latest changes. before doing it, i had no problem executing `-r all`, but after the upgrade i get the following panic message:

> thread '<unnamed>' panicked at 'internal error: entered unreachable code', /Users/foo/.cargo/registry/src/github.com-1ecc6299db9ec823/cargo-cache-0.3.2/src/cache/registry_pkg_cache.rs:122:17

when i re-run the command, the issue is gone.
fyi.

**Comment by matthiaskrgr** [comment]
Heh, I thought I just fixed that .... (the behaviour seems to be the one that I tried to fix in 24c58a3) :/

**Comment by matthiaskrgr** [comment]
@y0y0z oh I see, you are still on `0.3.2`....
I did not publish a new release yet, can you please try installing from git?:
`cargo install --git https://github.com/matthiaskrgr/cargo-cache --force` 

`cargo cache -V` should says something like `cargo-cache 0.3.2 (30f4bda 2019-09-22)` then.

>when i re-run the command, the issue is gone.
Yes I noticed this too, the panic occurred because I was initializing an in internal cache, removing a directory, and then partially updating the cache into an invalid state (which should be fixed now) which caused the panic.
The problem should be fixed with the latest git version I believe.

**Comment by y0y0z** [comment]
@matthiaskrgr: i can help you test it only if you will release a new version (i do not have a direct install access). but to be sure i am testing it as you expect, please describe exactly which steps should i take in order to reproduce the issue (or make sure that it is gone)

**Comment by y0y0z** [comment]
@matthiaskrgr: oopsy, closed the issue by mistake.

**Comment by matthiaskrgr** [comment]
Hi, I just released `0.3.3`, I hope everything works fine now! :)
If not, please let me know and open another ticket!
Cheers!

**Comment by y0y0z** [comment]
@matthiaskrgr: cannot reproduce the issue on the new version. seems everything works fine :)
